internal/client/db/db_test.go

```diff
@@ -0,0 +1,208 @@
+	db, err := newBlockchainDB[
+		hash.H256,
+		uint32,
+		runtime.BlakeTwo256,
+		noopExtrinsic,
+		*generic.Header[uint32, hash.H256, runtime.BlakeTwo256],
+	](database.NewMemDB[hash.H256]())
+	assert.NoError(t, err)
+	assert.NotNil(t, db)
+
+	db.clearPinningCache()
+}
+
+func TestBlockchainDB_insertJustificationsIfPinned(t *testing.T) {
+	db, err := newBlockchainDB[
+		hash.H256,
+		uint32,
+		runtime.BlakeTwo256,
+		noopExtrinsic,
+		*generic.Header[uint32, hash.H256, runtime.BlakeTwo256],
+	](database.NewMemDB[hash.H256]())
+	assert.NoError(t, err)
+	assert.NotNil(t, db)
+
+	someHash := hash.NewRandomH256()
+	someJustification := runtime.Justification{
+		ConsensusEngineID:    runtime.ConsensusEngineID{1, 1, 1, 1},
+		EncodedJustification: runtime.EncodedJustification{1, 1, 1, 1},
+	}
+	db.insertJustifcationsIfPinned(someHash, someJustification)
+	assert.False(t, db.pinnedBlocksCache.Contains(someHash))
+
+	db.pinnedBlocksCache.Pin(someHash)
+	db.insertJustifcationsIfPinned(someHash, someJustification)
+
+	assert.True(t, db.pinnedBlocksCache.Contains(someHash))
+}
+
+func TestBlockchainDB_insertPersistedJustificationsIfPinned(t *testing.T) {
+	memdb := database.NewMemDB[hash.H256]()
+	db, err := newBlockchainDB[
+		hash.H256,
+		uint32,
+		runtime.BlakeTwo256,
+		noopExtrinsic,
+		*generic.Header[uint32, hash.H256, runtime.BlakeTwo256],
+	](memdb)
+	assert.NoError(t, err)
+	assert.NotNil(t, db)
+
+	someHash := hash.NewRandomH256()
+	db.insertPersistedJustificationsIfPinned(someHash)
+	assert.False(t, db.pinnedBlocksCache.Contains(someHash))
+
+	// nothing in the db, but will pin `runtime.Justifications(nil)`
+	db.pinnedBlocksCache.Pin(someHash)
+	err = db.insertPersistedJustificationsIfPinned(someHash)
+	assert.NoError(t, err)
+	assert.True(t, db.pinnedBlocksCache.Contains(someHash))
+}
+
+func TestBlockchainDB_insertPersistedBodyIfPinned(t *testing.T) {
+	memdb := database.NewMemDB[hash.H256]()
+	db, err := newBlockchainDB[
+		hash.H256,
+		uint32,
+		runtime.BlakeTwo256,
+		noopExtrinsic,
+		*generic.Header[uint32, hash.H256, runtime.BlakeTwo256],
+	](memdb)
+	assert.NoError(t, err)
+	assert.NotNil(t, db)
+
+	someHash := hash.NewRandomH256()
+	db.insertPersistedBodyIfPinned(someHash)
+	assert.False(t, db.pinnedBlocksCache.Contains(someHash))
+
+	// nothing in the db, but will pin `[]runtime.Extrinsic(nil)`
+	db.pinnedBlocksCache.Pin(someHash)
+	err = db.insertPersistedBodyIfPinned(someHash)
+	assert.NoError(t, err)
+	assert.True(t, db.pinnedBlocksCache.Contains(someHash))
+}
+
+func TestBlockchainDB_bumpRef(t *testing.T) {
+	memdb := database.NewMemDB[hash.H256]()
+	db, err := newBlockchainDB[
+		hash.H256,
+		uint32,
+		runtime.BlakeTwo256,
+		noopExtrinsic,
+		*generic.Header[uint32, hash.H256, runtime.BlakeTwo256],
+	](memdb)
+	assert.NoError(t, err)
+	assert.NotNil(t, db)
+
+	someHash := hash.NewRandomH256()
+	db.bumpRef(someHash)
+	assert.True(t, db.pinnedBlocksCache.Contains(someHash))
+}
+
+func TestBlockchainDB_unpin(t *testing.T) {
```

Comment:
Do we want to add tests for some of the other public functions that are not tested here?

Reply:
As mentioned in the walkthrough,  the public methods will be tested when implementing the `db.Backend`.